File: c/grains/grains.c
#include "grains.h"

#include <math.h>

uint64_t square(uint8_t index) {
    uint64_t result = pow(2, index - 1);
    return result;
}

uint64_t total(void) {
    uint64_t result = 0;
    for(uint8_t index = 1; index <= 64; index++) {
        result += square(index);
    }
    return result;
}

File: scala/grains/src/main/scala/Grains.scala
import scala.math
import java.math.BigInteger

object Grains {
  def square(n: Int): Option[BigInt] = {
    if n > 64 || n < 1 then
      None
    else
      Some(BigInt(2).pow(n-1))
  }

  def total = {
    val exponents = Seq.range(1, 65)
    exponents.foldLeft(BigInt(0))((m, n) => m + square(n).get)
  }
}